### DISCUSSION TITLE
What are the differences between fingerprintjs (free) vs thumbarkjs? Which one is better?

### DISCUSSION THREAD

**Discussion by deadcoder0904** [thumbmarkjs/thumbmarkjs#15] [Q&A] [answered]
Curious, how it stacks against fingerprintjs as I don't know much about it so would be good to hear your thoughts.

Also, a comparison with creepjs would be great as well.

**Comment by ilkkapeltola** [comment, marked as answer]
FingerprintJS has covered more of the rare edge cases than what I have. ThumbmarkJS works equally well, or sometimes even better, in let's say 97% of the browsers, but FingerprintJS has better performance in the remaining 3%.

The reasons are pretty detailed. I'm able to actually use canvas fingerprinting in Samsung Internet, while FingerprintJS just disabled it.

Also ThumbmarkJS has a persistent fingerprint in Firefox normal vs. Incognito, while FingerprintJS does not.

CreepJS is not a library meant for the same thing. The code they have is excellent, very precise, elaborate etc. but it's heavy and not something you'd ever run on large scale.

They're proving how you are quite unique with a lot of techniques, but there's a lot there that makes it not production ready like this library. Just too heavy.

**Reply by deadcoder0904** [reply]
thanks, great answer!

**Reply by Valve** [reply]
@deadcoder0904 to clarify, the statement that FingerprintJS doesn't have a persistent fingerprint in Firefox normal vs incognito is not true.
You can verify it using the latest Firefox on the official FingerprintJS [demo page](https://fingerprintjs.github.io/fingerprintjs/).
Here is a screenshot of it showing the same fingerprint in normal vs incognito mode in the latest Firefox on a MacBook Air:
<img width="1470" alt="image" src="https://github.com/user-attachments/assets/2c2352dc-9e26-47b9-a742-2b74b8728a01" />

**Reply by ilkkapeltola** [reply]
A pretty old discussion. It was true back then. But seems to have changed for the new, non-free version.
The older version, the last truly free-to-use open source version 3.4.2 returns different hashes for normal and private FireFox windows.
But sure, the paid client side version has improved to the same standards as ThumbmarkJS now.

**Comment by jgabriel98** [comment]
All the links are invalid for me.
could anyone provide me an invitation for the discord?

**Comment by aleksa-codes** [comment]
I just want to say great work and thank you for the library, I faced an issue with FingerprintJS, where the generated id would be different on different zoom levels in Chrome, and switching to ThumbmarkJS solved it.

Also a bit off topic, I just want to point out a very funny and "unusual" thing. I have been running into @deadcoder0904 so many times on different platforms, for very specific things and questions, it is kinda insane. We either have very similar taste for tech stacks, libraries, run into same problems all the time or I don't know what the deal is 😂 It has happened so many times it can't be a coincidence anymore lol

**Reply by deadcoder0904** [reply]
I'm everywhere.

**Comment by smhz101** [comment]
I’ve used both recently, and one thing I noticed is that ThumbmarkJS seems more stable in some cases (especially browser zoom/rendering changes), while FingerprintJS feels more polished in ecosystem/docs and easier to integrate quickly.

I think the "better" choice depends on the use case as always:
- Fraud detection / production SaaS scale: FingerprintJS (especially paid versions) has more maturity and infrastructure behind it.
- Open-source / self-hosted / lightweight control: ThumbmarkJS looks very compelling.

Right now it feels like ThumbmarkJS is the strongest open-source alternative I’ve seen.
